arrow-json/src/writer/encoder.rs

```diff
@@ -0,0 +1,436 @@
+            (Box::new(MapEncoder::try_new(array, options)?) as _,  array.nulls().cloned())
+        }
+
+        DataType::Struct(fields) => {
+            let array = array.as_struct();
+            let encoders = fields.iter().zip(array.columns()).map(|(field, array)| {
+                let (encoder, nulls) = make_encoder_impl(array, options)?;
+                Ok(FieldEncoder{
+                    field: field.clone(),
+                    encoder, nulls
+                })
+            }).collect::<Result<Vec<_>, ArrowError>>()?;
+
+            let encoder = StructArrayEncoder{
+                encoders,
+                explicit_nulls: options.explicit_nulls,
+            };
+            (Box::new(encoder) as _, array.nulls().cloned())
+        }
+        d => match d.is_temporal() {
+            true => {
+                // Note: the implementation of Encoder for ArrayFormatter assumes it does not produce
+                // characters that would need to be escaped within a JSON string, e.g. `'"'`.
+                // If support for user-provided format specifications is added, this assumption
+                // may need to be revisited
+                let options = FormatOptions::new().with_display_error(true);
+                let formatter = ArrayFormatter::try_new(array, &options)?;
+                (Box::new(formatter) as _, array.nulls().cloned())
+            }
+            false => return Err(ArrowError::InvalidArgumentError(format!("JSON Writer does not support data type: {d}"))),
+        }
+    })
+}
+
+fn encode_string(s: &str, out: &mut Vec<u8>) {
+    let mut serializer = serde_json::Serializer::new(out);
+    serializer.serialize_str(s).unwrap();
+}
+
+struct FieldEncoder<'a> {
+    field: FieldRef,
+    encoder: Box<dyn Encoder + 'a>,
+    nulls: Option<NullBuffer>,
+}
+
+struct StructArrayEncoder<'a> {
+    encoders: Vec<FieldEncoder<'a>>,
+    explicit_nulls: bool,
+}
+
+impl<'a> Encoder for StructArrayEncoder<'a> {
+    fn encode(&mut self, idx: usize, out: &mut Vec<u8>) {
+        out.push(b'{');
+        let mut is_first = true;
+        for field_encoder in &mut self.encoders {
+            let is_null = field_encoder.nulls.as_ref().is_some_and(|n| n.is_null(idx));
+            if is_null && !self.explicit_nulls {
+                continue;
+            }
+
+            if !is_first {
+                out.push(b',');
+            }
+            is_first = false;
+
+            encode_string(field_encoder.field.name(), out);
+            out.push(b':');
+
+            match is_null {
+                true => out.extend_from_slice(b"null"),
+                false => field_encoder.encoder.encode(idx, out),
+            }
+        }
+        out.push(b'}');
+    }
+}
+
+trait PrimitiveEncode: ArrowNativeType {
+    type Buffer;
+
+    // Workaround https://github.com/rust-lang/rust/issues/61415
+    fn init_buffer() -> Self::Buffer;
+
+    fn encode(self, buf: &mut Self::Buffer) -> &[u8];
```

Comment:
I think it would help to document what encode does here
```suggestion
    /// Encode the primitive value as bytes, returning a reference to that slice. 
    /// `buf` is temporary space that may be used
    fn encode(self, buf: &mut Self::Buffer) -> &[u8];
```